test/common/network/utility_test.cc

```diff
@@ -125,7 +128,122 @@ TEST_P(NetworkUtilityGetLocalAddress, GetLocalAddress) {
   EXPECT_NE(nullptr, Utility::getLocalAddress(GetParam()));
 }
 
-TEST(NetworkUtility, GetOriginalDst) { EXPECT_EQ(nullptr, Utility::getOriginalDst(-1)); }
+TEST(NetworkUtility, GetOriginalDstBadFd) { EXPECT_EQ(nullptr, Utility::getOriginalDst(-1)); }
+
+class NetworkUtilityGetOriginalDst : public testing::TestWithParam<Address::IpVersion> {
+public:
+  NetworkUtilityGetOriginalDst() {
+    EXPECT_EQ(pthread_mutex_init(&mutex_, nullptr), 0);
+    EXPECT_EQ(pthread_cond_init(&cond_, nullptr), 0);
+  }
+
+  ~NetworkUtilityGetOriginalDst() {
+    EXPECT_EQ(pthread_cond_destroy(&cond_), 0);
+    EXPECT_EQ(pthread_mutex_destroy(&mutex_), 0);
+  }
+
+  void makeFdBlocking(int fd) {
+    const int flags = ::fcntl(fd, F_GETFL, 0);
+    ASSERT_GE(flags, 0);
+    ASSERT_EQ(::fcntl(fd, F_SETFL, flags & (~O_NONBLOCK)), 0);
+  }
+
+  void wait() {
+    EXPECT_EQ(pthread_mutex_lock(&mutex_), 0);
+
+    int rc = 0;
+    while(rc == 0 && !signaled_) {
+      rc = pthread_cond_wait(&cond_, &mutex_);
+    }
+
+    EXPECT_EQ(pthread_mutex_unlock(&mutex_), 0);
+    EXPECT_EQ(rc, 0);
+  }
+
+  void signal() {
+    EXPECT_EQ(pthread_mutex_lock(&mutex_), 0);
+    signaled_ = 1;
+    EXPECT_EQ(pthread_cond_signal(&cond_), 0);
+    EXPECT_EQ(pthread_mutex_unlock(&mutex_), 0);
+  }
+
+protected:
+  pthread_mutex_t mutex_;
+  pthread_cond_t cond_;
+  bool signaled_ = 0;
+};
+
+INSTANTIATE_TEST_CASE_P(IpVersions, NetworkUtilityGetOriginalDst,
+                        testing::ValuesIn(TestEnvironment::getIpVersionsForTest()));
+
+TEST_P(NetworkUtilityGetOriginalDst, GetOriginalDst) {
```

Comment:
I agree. I would just revert all of this IMO and only do the SOL_IP call on linux, and on apple just set error to ENOTSUPPORTED (or whatever) so that the subsequence if check fails with a TODO.